contracts/strategies/allocation/AllocationWithVotingBase.sol

```diff
@@ -0,0 +1,306 @@
+                recipient: recipient,
+                amount: 0,
+                votes: 0,
+                status: Status.PENDING,
+                metadata: metadata,
+                creator: _sender
+            });
+
+            applications[applicationId] = application;
+        } else {
+            Application storage application;
+            application = applications[applicationId];
+
+            if (identityRequired && !registry.isOwnerOrMemberOfIdentity(application.identityId, _sender)) {
+                /// used in allocation strategy where identityRequired is true
+                revert UNAUTHORIZED();
+            } else if (application.creator != _sender) {
+                /// used in allocation strategy where identityRequired is false
+                revert UNAUTHORIZED();
+            }
+
+            // Update the application
+            application.recipient = recipient;
+            application.status = Status.REAPPLIED;
+            application.metadata = metadata;
+
+            applications[applicationId] = application;
+        }
+
+        // Add the application to the applications mapping
+        // Note: can't do it here.  Need to do it in the if/else above.
+        // applications[applicationId] = application;
+        applicationStatus[applicationId] = IAllocationStrategy.ApplicationStatus.Pending;
+
+        emit ApplicationSubmitted(applicationId, identityId, recipient, metadata, _sender);
+
+        return applicationId;
+    }
+
+    /// @notice Returns the status of the application
+    /// @param applicationId The applicationId of the application
+    function getApplicationStatus(uint256 applicationId) external view override returns (ApplicationStatus) {
+        return applicationStatus[applicationId];
+    }
+
+    /// @notice Allocates votes to the application(s)
+    /// @param _data The data to be decoded
+    /// @param sender The sender of the transaction
+    function allocate(bytes memory _data, address sender) external payable override onlyAllo returns (uint256) {
+        if (votingStartTime > block.timestamp || votingEndTime < block.timestamp) {
+            revert VOTING_NOT_OPEN();
+        }
+
+        // Decode the _data
+        // Note: should have option to vote on multiple applications so an array of Payouts is expected.
+        (Transfer.Payout[] memory payouts, address token, uint256[] memory applicationIds) =
+            abi.decode(_data, (Transfer.Payout[], address, uint256[]));
+
+        if (payouts.length > 1) {
+            // Transfer funds
+            // Note: this should live in distribution strategy
+            // bool transferSuccess = Transfer.transferPayouts(token, payouts);
+            // if (!transferSuccess) {
+            //     revert Transfer.TRANSFER_FAILED();
+            // }
+
+            // Loop through the payouts and update the votes on each application
+            for (uint256 i; i < payouts.length; i++) {
```

Comment:
i think there is a problem in the logic

for **every** receipient in payouts we iterate through **all** applicationIds and allocate the amounts
my suggestion: we shouldnt use `Payout` here or we should change payout to
```
Payout {
   uint256 applicationId;
   uint256 amount;
}
```

in general i would suggest a format like:

```
struct SingleAlloction {
   uint256 applicationId;
   uint256 amount;
}

struct Allocation {
   address token;
   uint256 totalAmount;
   SingleAllocation[] allocations;
}
```

for every `token` in the list, we would have one transfer of `totalAmount` and than we could iterate over the `allocations` and add the amounts.
if we do not allow to allocate different token we could remove the outer struct.
if we go with multiple token, we need to introduce a `token -> amount` mapping for each user

from spec:
> donations can be sent in native tokens or ERC20s

Reply:
No we will allowing only 1 token per transaction.
That token can be an ERC20 / native token